File: python/pyautd3/gain/focus.py
"""
File: focus.py
Project: gain
Created Date: 14/09/2023
Author: Shun Suzuki
-----
Last Modified: 02/10/2023
Modified By: Shun Suzuki (suzuki@hapis.k.u-tokyo.ac.jp)
-----
Copyright (c) 2023 Shun Suzuki. All rights reserved.

"""


import numpy as np
from numpy.typing import ArrayLike

from pyautd3.emit_intensity import EmitIntensity
from pyautd3.geometry import Geometry
from pyautd3.internal.gain import IGain
from pyautd3.native_methods.autd3capi import NativeMethods as Base
from pyautd3.native_methods.autd3capi_def import GainPtr


class Focus(IGain):
    """Gain to produce a focal point."""

    _p: np.ndarray
    _intensity: EmitIntensity | None

    def __init__(self: "Focus", pos: ArrayLike) -> None:
        """Constructor.

        Arguments:
        ---------
            pos: Position of the focal point
        """
        super().__init__()
        self._p = np.array(pos)
        self._intensity = None

    def with_intensity(self: "Focus", intensity: int | EmitIntensity) -> "Focus":
        """Set amplitude.

        Arguments:
        ---------
            intensity: Emission intensity
        """
        self._intensity = EmitIntensity._cast(intensity)
        return self

    def _gain_ptr(self: "Focus", _: Geometry) -> GainPtr:
        ptr = Base().gain_focus(self._p[0], self._p[1], self._p[2])
        if self._intensity is not None:
            ptr = Base().gain_focus_with_intensity(ptr, self._intensity.value)
        return ptr

File: dotnet/cs/src/Gain/Focus.cs
/*
 * File: Focus.cs
 * Project: Gain
 * Created Date: 13/09/2023
 * Author: Shun Suzuki
 * -----
 * Last Modified: 29/11/2023
 * Modified By: Shun Suzuki (suzuki@hapis.k.u-tokyo.ac.jp)
 * -----
 * Copyright (c) 2023 Shun Suzuki. All rights reserved.
 * 
 */

#if UNITY_2018_3_OR_NEWER
#define USE_SINGLE
#endif

#if UNITY_2020_2_OR_NEWER
#nullable enable
#endif

using AUTD3Sharp.NativeMethods;

#if UNITY_2018_3_OR_NEWER
using UnityEngine;
using Vector3 = UnityEngine.Vector3;
#else
using Vector3 = AUTD3Sharp.Utils.Vector3d;
#endif

#if USE_SINGLE
using float_t = System.Single;
#else
using float_t = System.Double;
#endif

namespace AUTD3Sharp.Gain
{
    /// <summary>
    /// Gain to produce single focal point
    /// </summary>
    public sealed class Focus : Internal.Gain
    {
        private readonly Vector3 _point;
        private EmitIntensity? _intensity;

        public Focus(Vector3 point)
        {
            _point = point;
            _intensity = null;
        }

        /// <summary>
        /// Set amplitude
        /// </summary>
        /// <param name="intensity">Emission intensity</param>
        /// <returns></returns>
        public Focus WithIntensity(byte intensity)
        {
            _intensity = new EmitIntensity(intensity);
            return this;
        }

        /// <summary>
        /// Set amplitude
        /// </summary>
        /// <param name="intensity">Emission intensity</param>
        /// <returns></returns>
        public Focus WithIntensity(EmitIntensity intensity)
        {
            _intensity = intensity;
            return this;
        }

        internal override GainPtr GainPtr(Geometry geometry)
        {
            var ptr = NativeMethodsBase.AUTDGainFocus(_point.x, _point.y, _point.z);
            if (_intensity != null)
                ptr = NativeMethodsBase.AUTDGainFocusWithIntensity(ptr, _intensity.Value.Value);
            return ptr;
        }
    }
}

#if UNITY_2020_2_OR_NEWER
#nullable restore
#endif
